plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Transform/CaseOfCase.hs

```diff
@@ -0,0 +1,34 @@
+{-# LANGUAGE LambdaCase    #-}
+{-# LANGUAGE TypeFamilies  #-}
+{-# LANGUAGE TypeOperators #-}
+
+{- | A special case of case-of-case optimisation: transforms
+
+@
+    case ((force ifThenElse) b (constr t) (constr f)) alts
+@
+
+into
+
+@
+    force ifThenElse b (case (constr t) branches) (case (constr f) branches)
+@
+
+This is always an improvement.
+-}
+module UntypedPlutusCore.Transform.CaseOfCase (caseOfCase) where
+
+import PlutusCore qualified as PLC
+import UntypedPlutusCore.Core
+
+import Control.Lens
+
+caseOfCase :: (fun ~ PLC.DefaultFun) => Term name uni fun a -> Term name uni fun a
+caseOfCase = transformOf termSubterms $ \case
+  Case ann scrut alts
+    | ( ite@(Force _ (Builtin _ PLC.IfThenElse))
+        , [cond, (true@Constr{}, trueAnn), (false@Constr{}, falseAnn)]
```

Comment:
Wait, so it works for `ifThenElse cond nil (cons x xs)` too (with `nil` and `cons` being SOP-based)? I guess it's even better, but is worth explicit discussion and the naming probably shouldn't be limited to `true` and `false`, as well as the specification above.

Reply:
> The name `true/false` is because those are the true and false branches of `ifThenElse`.

I just realized that they are in fact not, in the general and perfectly realistic case. Think of how `not (if b then True else False) && c` or whatever would look like after `not` is optimized, but before `&&` is optimized (assuming both get inlined).

Also could you please add a test showing how `not (not (if b then True else False))` gets optimized (with `not` inlined)?

Reply:
This PR is not about optimizing `not . not`. I think that requires the actual case-of-case optimization. This PR implements a special case of case-of-case, where the inner case is `(builtin ifThenElse)`. This special case occurs very frequently, namely every time one writes something similar to `if x < y then ... else ...`. `not . not` is not covered by this special case.

Reply:
Added a test case involving `not . not` (`compiledNotNot`), and it currently isn't fully optimized (but that's not the goal of this PR). See `not-not.uplc-readable.golden`.

Reply:
I think you only this optimization and case-of-known-constructor? Here's how interleaving those two would transform the `not-not` expression:

```
   not (not (if b then True else False))
~> not (if b then not True else not False)
~> not (if b then False else True)
~> if b then not False else not True
~> if b then True else False
```

Reply:
(that wasn't a response to your last message, which I hadn't seen up until after I wrote my comment)

Reply:
>  it currently isn't fully optimized (but that's not the goal of this PR). See `not-not.uplc-readable.golden`.

Look like we should propagate `force` the same way this PR propagates `case`?

Reply:
> Look like we should propagate force the same way this PR propagates case?

I think we can do that - as long as all the `delay` branches are pure. That is a separate optimization though.